### PR TITLE
guide: add a testing section to the contributing guide

### PR DESCRIPTION
**Pull request by carllerche**
## Motivation

We would like contributors to be able to help out with documentation tests

## Solution

Add some notes in the guide on how to write tests.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -112,6 +112,104 @@ usually a good idea to first open an issue describing the change to solicit
...
+to use in tests, no matter the crate being tested.
+
+The best strategy for writing a new integration test is to look at existing
+integration tests in the crate and following the style.
+
+#### Documentation tests
+
+Ideally, every API has at least one documentation test that demonstrates how to
+use the API. Documentation tests are run with `cargo test --doc`. This ensures
+that the example is correct and provides additional test coverage.
```
Comment (hawkw) [inline CONTRIBUTING.md]: TIOLI: might be good to link to https://doc.rust-lang.org/rustdoc/documentation-tests.html somewhere in here?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -112,6 +112,104 @@ usually a good idea to first open an issue describing the change to solicit
...
+/// process.timeout(Duration::from_millis(10))
+/// # ).unwrap();
+/// # }
+```
+
+Given that this is a *type* level documentation test and the primary way users
+of `tokio` will create an instance of `Timeout` is by using
+`FutureExt::timeout`, this is how the documentation test is structured.
+
+Lines that start with `/// #` are removed when the documentation is generated.
```
Comment (hawkw) [inline CONTRIBUTING.md]: Can we add a note in here some place that the minimum supported Rust version is 1.25, and on that version, any test which uses `extern crate` but has no `fn main` needs an empty `fn main` somewhere, or else the `extern crate` will not work correctly? See commit 4704be05fb8e904e91c8b725f973749ff49776df and rust-lang/rust#48106.

Comment (carllerche) [reply CONTRIBUTING.md]: Starting at which version is the main fn no longer needed?

Comment (hawkw) [reply CONTRIBUTING.md]: 1.26+

Comment (hawkw) [reply CONTRIBUTING.md]: This is resolved if we're bumping the min version to 1.26.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -112,6 +112,104 @@ usually a good idea to first open an issue describing the change to solicit
...
+The best strategy for writing a new integration test is to look at existing
+integration tests in the crate and following the style.
+
+#### Documentation tests
+
+Ideally, every API has at least one [documentation test] that demonstrates how to
+use the API. Documentation tests are run with `cargo test --doc`. This ensures
+that the example is correct and provides additional test coverage.
+
+The trick to documentation tests is stricking a balance between being susinct
```
Comment (tobz) [inline CONTRIBUTING.md]: nit: `stricking` -> `striking`, `sucinct` -> `succinct`

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -112,6 +112,104 @@ usually a good idea to first open an issue describing the change to solicit
...
+/// # ).unwrap();
+/// # }
+```
+
+Given that this is a *type* level documentation test and the primary way users
+of `tokio` will create an instance of `Timeout` is by using
+`FutureExt::timeout`, this is how the documentation test is structured.
+
+Lines that start with `/// #` are removed when the documentation is generated.
+They are onlly there to get the test to run. The `block_on_all` function is the
```
Comment (tobz) [inline CONTRIBUTING.md]: nit: `onlly` -> `only`

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -112,6 +112,104 @@ usually a good idea to first open an issue describing the change to solicit
...
+If the change being proposed alters code (as opposed to only documentation for
+example), it is either adding new functionality to Tokio or it is fixing
+existing, broken functionality. In both of these cases, the pull request should
+include one or more tests to ensure that Tokio does not regress in the future.
+There are two ways to write tests: integration tests and documentation tests
+(Tokio avoids unit tests as much as possible).
+
+#### Integration tests
+
+Integration tests go in the crate that they are test. Each sub crate should have
```
Comment (tobz) [inline CONTRIBUTING.md]: Reword this slightly? `Integration tests go in the same crate as the code they are testing`, perhaps?

**Pull request merged**